File: fm-domain/src/main/java/io/zeitwert/fm/util/SqlUtils.java
package io.zeitwert.fm.util;

import static io.dddrive.util.Invariant.assertThis;

import java.time.Instant;
import java.time.LocalDateTime;
import java.time.OffsetDateTime;
import java.time.ZoneId;
import java.time.ZoneOffset;
import java.util.Collection;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

import org.jooq.Condition;
import org.jooq.Field;
import org.jooq.SortField;
import org.jooq.SortOrder;
import org.jooq.Table;
import org.jooq.impl.DSL;

import io.crnk.core.queryspec.Direction;
import io.crnk.core.queryspec.FilterOperator;
import io.crnk.core.queryspec.FilterSpec;
import io.crnk.core.queryspec.SortSpec;

public class SqlUtils {

	public interface SearchConditionProvider {
		Condition apply(Field<Integer> idField, FilterSpec filter);
	}

	private final SearchConditionProvider searchConditionProvider;

	public SqlUtils(SearchConditionProvider searchConditionProvider) {
		this.searchConditionProvider = searchConditionProvider;
	}

	public Condition andFilter(Condition whereClause, Table<?> table, Field<Integer> idField, FilterSpec filter) {
		return whereClause.and(this.filter(table, idField, filter));
	}

	public Condition orFilter(Condition whereClause, Table<?> table, Field<Integer> idField, FilterSpec filter) {
		List<Condition> conditions = filter.getExpression().stream()
				.map(f -> this.filter(table, idField, f)).toList();
		return whereClause.and(DSL.or(conditions));
	}

	public List<SortField<?>> sortFilter(Table<?> table, List<SortSpec> sortSpec) {
		return sortSpec.stream()
				.map(s -> {
					return table.field(StringUtils.toSnakeCase(s.getPath().toString()))
							.sort(Direction.ASC.equals(s.getDirection()) ? SortOrder.ASC : SortOrder.DESC).nullsLast();
				}).collect(Collectors.toList());
	}

	private Condition closedFilter(Table<?> table, FilterSpec filter) {
		Field<?> field = table.field("closed_at");
		assertThis(field != null, "known field closed_at");
		if (field == null) {
			return null; // make compiler happy (potential null pointer)
		}
		Boolean value = filter.getValue();
		if (value) {
			if (filter.getOperator() == FilterOperator.EQ) {
				return field.isNotNull();
			} else if (filter.getOperator() == FilterOperator.NEQ) {
				return field.isNull();
			}
		} else if (!value) {
			if (filter.getOperator() == FilterOperator.EQ) {
				return field.isNull();
			} else if (filter.getOperator() == FilterOperator.NEQ) {
				return field.isNotNull();
			}
		}
		return DSL.trueCondition();
	}

	private Integer toInteger(Object value) {
		if (value == null) {
			return null;
		} else if (value.getClass().isArray() && ((Object[]) value).length == 1) {
			return this.toInteger(((Object[]) value)[0]);
		} else if (value instanceof Collection<?> && ((Collection<?>) value).size() == 1) {
			return this.toInteger(((Collection<?>) value).stream().toList().get(0));
		} else if (value.getClass() == Integer.class) {
			return (Integer) value;
		} else if (value.getClass() == String.class) {
			return Integer.valueOf((String) value);
		}
		assertThis(false, value + " (" + value.getClass() + ") is an integer");
		return null;
	}

	@SuppressWarnings("unchecked")
	private Condition integerFilter(Field<Integer> field, FilterSpec filter) {
		if (filter.getValue() instanceof Collection) {
			if (filter.getOperator() == CustomFilters.IN || filter.getOperator() == FilterOperator.EQ) {
				return field.in((Collection<Integer>) filter.getValue());
			}
			assertThis(false, "supported integer filter operator " + filter.getOperator() + " on " + filter.getValue());
		} else {
			Integer value = this.toInteger(filter.getValue());
			if (filter.getOperator() == FilterOperator.EQ) {
				if (value != null) {
					return field.eq(value);
				} else {
					return field.isNull();
				}
			} else if (filter.getOperator() == FilterOperator.NEQ) {
				if (value != null) {
					return field.eq(value).not();
				} else {
					return field.isNotNull();
				}
			} else if (filter.getOperator() == FilterOperator.GT) {
				return field.gt(value);
			} else if (filter.getOperator() == FilterOperator.GE) {
				return field.ge(value);
			} else if (filter.getOperator() == FilterOperator.LT) {
				return field.lt(value);
			} else if (filter.getOperator() == FilterOperator.LE) {
				return field.le(value);
			}
		}
		assertThis(false, "supported integer filter operator " + filter.getOperator() + " on " + filter.getValue());
		return DSL.falseCondition();
	}

	private Condition stringFilter(Field<String> field, FilterSpec filter) {
		if (filter.getOperator() == CustomFilters.IN) {
			Set<String> value = filter.getValue();
			Condition inner = DSL.noCondition();
			for (String val : value) {
				inner = inner.or(field.eq(val));
			}
			return inner;
		}

		String value = filter.getValue() != null ? filter.getValue().toString() : null;
		if (filter.getOperator() == FilterOperator.EQ) {
			if (value != null) {
				return field.eq(value);
			} else {
				return field.isNull().or(field.eq(""));
			}
		} else if (filter.getOperator() == FilterOperator.NEQ) {
			if (value != null) {
				return field.eq(value).not();
			} else {
				return field.isNotNull().and(field.eq("").not());
			}
		} else if (filter.getOperator() == FilterOperator.GT) {
			return field.gt(value);
		} else if (filter.getOperator() == FilterOperator.GE) {
			return field.ge(value);
		} else if (filter.getOperator() == FilterOperator.LT) {
			return field.lt(value);
		} else if (filter.getOperator() == FilterOperator.LE) {
			return field.le(value);
		} else if (filter.getOperator() == FilterOperator.LIKE && value != null) {
			return DSL.lower(field).like(value.replace("*", "%"));
		}

		assertThis(false, "supported string filter operator " + filter.getOperator());
		return DSL.falseCondition();
	}

	private Condition booleanFilter(Field<Boolean> field, FilterSpec filter) {
		Boolean value = filter.getValue();
		if (filter.getOperator() == FilterOperator.EQ) {
			return field.eq(value);
		}
		assertThis(false, "supported boolean filter operator " + filter.getOperator());
		return DSL.falseCondition();
	}

	private Condition localDateTimeFilter(Field<LocalDateTime> field, FilterSpec filter) {
		LocalDateTime value = filter.getValue();
		if (filter.getOperator() == FilterOperator.EQ) {
			if (value != null) {
				return field.eq(value);
			} else {
				return field.isNull();
			}
		} else if (filter.getOperator() == FilterOperator.NEQ) {
			if (value != null) {
				return field.eq(value).not();
			} else {
				return field.isNotNull();
			}
		} else if (filter.getOperator() == FilterOperator.GT) {
			return field.gt(value);
		} else if (filter.getOperator() == FilterOperator.GE) {
			return field.ge(value);
		} else if (filter.getOperator() == FilterOperator.LT) {
			return field.lt(value);
		} else if (filter.getOperator() == FilterOperator.LE) {
			return field.le(value);
		}
		assertThis(false, "supported local date time filter operator " + filter.getOperator());
		return DSL.falseCondition();
	}

	private Condition offsetDateTimeFilter(Field<OffsetDateTime> field, FilterSpec filter) {
		OffsetDateTime value = (OffsetDateTime) filter.getValue();
		if (value == null) {
			if (filter.getOperator() == FilterOperator.EQ) {
				return field.isNull();
			} else if (filter.getOperator() == FilterOperator.NEQ) {
				return field.isNotNull();
			}
		} else {
			ZoneOffset zoneOffset = ZoneId.systemDefault().getRules().getOffset(Instant.now());
			value = value.atZoneSameInstant(zoneOffset).toOffsetDateTime();
			if (filter.getOperator() == FilterOperator.EQ) {
				return field.eq(value);
			} else if (filter.getOperator() == FilterOperator.NEQ) {
				return field.eq(value).not();
			} else if (filter.getOperator() == FilterOperator.GT) {
				return field.gt(value);
			} else if (filter.getOperator() == FilterOperator.GE) {
				return field.ge(value);
			} else if (filter.getOperator() == FilterOperator.LT) {
				return field.lt(value);
			} else if (filter.getOperator() == FilterOperator.LE) {
				return field.le(value);
			}
		}
		assertThis(false, "supported offset date time filter operator " + filter.getOperator());
		return DSL.falseCondition();
	}

	@SuppressWarnings("unchecked")
	private Condition filter(Table<?> table, Field<Integer> idField, FilterSpec filter) {
		String fieldName = StringUtils.toSnakeCase(CrnkUtils.getPath(filter));
		if ("search_text".equals(fieldName)) {
			return this.searchConditionProvider.apply(idField, filter);
		} else if ("is_closed".equals(fieldName)) {
			return this.closedFilter(table, filter);
		}
		Field<?> field = table.field(fieldName);
		assertThis(field != null, "known field " + fieldName);
		if (field == null) {
			return null; // make compiler happy (potential null pointer)
		}
		if (field.getType() == Integer.class) {
			return this.integerFilter((Field<Integer>) field, filter);
		} else if (field.getType() == String.class) {
			return this.stringFilter((Field<String>) field, filter);
		} else if (field.getType() == Boolean.class) {
			return this.booleanFilter((Field<Boolean>) field, filter);
		} else if (field.getType() == LocalDateTime.class) {
			return this.localDateTimeFilter((Field<LocalDateTime>) field, filter);
		} else if (field.getType() == OffsetDateTime.class) {
			return this.offsetDateTimeFilter((Field<OffsetDateTime>) field, filter);
		}
		assertThis(false, "supported field type " + fieldName + ": " + field.getType());
		return DSL.falseCondition();
	}

}

File: fm-domain/src/main/java/io/zeitwert/dddrive/persist/util/SqlUtils.kt
package io.zeitwert.dddrive.persist.util

import io.crnk.core.queryspec.Direction
import io.crnk.core.queryspec.FilterOperator
import io.crnk.core.queryspec.FilterSpec
import io.crnk.core.queryspec.SortSpec
import io.zeitwert.dddrive.persist.util.CrnkUtils.getPath
import io.zeitwert.fm.obj.model.db.Tables
import org.jooq.Condition
import org.jooq.Field
import org.jooq.SortOrder
import org.jooq.Table
import org.jooq.impl.DSL
import org.slf4j.LoggerFactory
import java.time.Instant
import java.time.LocalDateTime
import java.time.OffsetDateTime
import java.time.ZoneId

class SqlUtils {

	companion object {

		val logger = LoggerFactory.getLogger(SqlUtils::class.java)!!
	}

	private val searchConditionProvider: SearchConditionProvider?

	constructor() {
		this.searchConditionProvider = null
	}

	constructor(searchConditionProvider: SearchConditionProvider?) {
		this.searchConditionProvider = searchConditionProvider
	}

	fun andFilter(
		whereClause: Condition,
		table: Table<*>,
		idField: Field<Int>,
		filter: FilterSpec,
	) = whereClause.and(filter(table, idField, filter))

	fun orFilter(
		whereClause: Condition,
		table: Table<*>,
		idField: Field<Int>,
		filter: FilterSpec,
	) = whereClause.and(DSL.or(filter.expression.map { filter(table, idField, it) }))

	fun sortFields(
		table: Table<*>,
		sortSpec: List<SortSpec>,
	) = sortSpec.map {
		table
			.field(StringUtils.toSnakeCase(it.path.toString()))!!
			.sort(if (Direction.ASC == it.direction) SortOrder.ASC else SortOrder.DESC)
			.nullsLast()
	}

	private fun closedFilter(
		table: Table<*>,
		filter: FilterSpec,
	): Condition {
		val field = table.field(Tables.OBJ.CLOSED_AT.name)!!
		val value = filter.getValue<Boolean>()
		if (value) {
			if (filter.operator === FilterOperator.EQ) {
				return field.isNotNull()
			} else if (filter.operator === FilterOperator.NEQ) {
				return field.isNull()
			}
		} else if (!value) {
			if (filter.operator === FilterOperator.EQ) {
				return field.isNull()
			} else if (filter.operator === FilterOperator.NEQ) {
				return field.isNotNull()
			}
		}
		return DSL.trueCondition()
	}

	@Suppress("UNCHECKED_CAST")
	private fun toInteger(value: Any?): Int? =
		if (value == null) {
			null
		} else if (value.javaClass.isArray && (value as Array<Any>).size == 1) {
			toInteger(value[0])
		} else if (value is Collection<*> && value.size == 1) {
			toInteger(value.first())
		} else if (value.javaClass == Int::class.java || value.javaClass == Integer::class.java) {
			value as Int
		} else if (value.javaClass == String::class.java || value.javaClass == java.lang.String::class.java) {
			(value as String).toInt()
		} else {
			throw IllegalArgumentException("" + value + " (" + value.javaClass + ") is not an integer")
		}

	@Suppress("UNCHECKED_CAST")
	private fun integerFilter(
		field: Field<Int>,
		filter: FilterSpec,
	): Condition {
		val fieldName = StringUtils.toSnakeCase(getPath(filter))
		val fieldValue = filter.getValue<Any?>()
		logger.trace("integerFilter({} {} {} ({})", fieldName, filter.operator, fieldName, fieldValue?.javaClass)
		try {
			return if (fieldValue is Collection<*>) {
				if (filter.operator === CustomFilters.IN || filter.operator === FilterOperator.EQ) {
					return field.`in`(fieldValue as Collection<Int>?)
				} else {
					throw IllegalArgumentException("Unsupported integer collection operator $fieldName ${filter.operator} $fieldValue (${fieldValue.javaClass})")
				}
			} else {
				val intValue = toInteger(fieldValue)
				if (filter.operator === FilterOperator.EQ) {
					if (intValue != null) {
						field.eq(intValue)
					} else {
						field.isNull()
					}
				} else if (filter.operator === FilterOperator.NEQ) {
					if (intValue != null) {
						field.eq(intValue).not()
					} else {
						field.isNotNull()
					}
				} else if (filter.operator === FilterOperator.GT) {
					field.gt(intValue)
				} else if (filter.operator === FilterOperator.GE) {
					field.ge(intValue)
				} else if (filter.operator === FilterOperator.LT) {
					field.lt(intValue)
				} else if (filter.operator === FilterOperator.LE) {
					field.le(intValue)
				} else {
					throw IllegalArgumentException("Unsupported integer operator $fieldName ${filter.operator} $intValue (${intValue?.javaClass})")
				}
			}
		} catch (e: Exception) {
			throw RuntimeException(
				"integerFilter $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass}) crashed: ${e.message}",
				e,
			)
		}
	}

	private fun stringFilter(
		field: Field<String>,
		filter: FilterSpec,
	): Condition {
		if (filter.operator === CustomFilters.IN) {
			val value = filter.getValue<Set<String>>()
			var inner = DSL.noCondition()
			for (`val` in value) {
				inner = inner.or(field.eq(`val`))
			}
			return inner
		}
		val fieldName = StringUtils.toSnakeCase(getPath(filter))
		val fieldValue = filter.getValue<Any?>()?.toString()
		try {
			return if (filter.operator === FilterOperator.EQ) {
				if (fieldValue != null) {
					field.eq(fieldValue)
				} else {
					field.isNull().or(field.eq(""))
				}
			} else if (filter.operator === FilterOperator.NEQ) {
				if (fieldValue != null) {
					field.eq(fieldValue).not()
				} else {
					field.isNotNull().and(field.eq("").not())
				}
			} else if (filter.operator === FilterOperator.GT) {
				field.gt(fieldValue)
			} else if (filter.operator === FilterOperator.GE) {
				field.ge(fieldValue)
			} else if (filter.operator === FilterOperator.LT) {
				field.lt(fieldValue)
			} else if (filter.operator === FilterOperator.LE) {
				field.le(fieldValue)
			} else if (filter.operator === FilterOperator.LIKE && fieldValue != null) {
				DSL.lower(field).like(fieldValue.replace("*", "%"))
			} else {
				throw IllegalArgumentException("Unsupported string operator $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass})")
			}
		} catch (e: Exception) {
			throw RuntimeException(
				"stringFilter $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass}) crashed: ${e.message}",
				e,
			)
		}
	}

	private fun booleanFilter(
		field: Field<Boolean>,
		filter: FilterSpec,
	): Condition {
		val fieldName = StringUtils.toSnakeCase(getPath(filter))
		val anyValue = filter.getValue<Any?>()
		val fieldValue = when (anyValue) {
			is Boolean -> anyValue
			is String -> anyValue.toBoolean()
			else -> throw IllegalArgumentException("Unsupported boolean value type for $fieldName: $anyValue (${anyValue?.javaClass})")
		}
		try {
			return if (filter.operator === FilterOperator.EQ) {
				field.eq(fieldValue)
			} else {
				throw IllegalArgumentException("Unsupported boolean operator $fieldName ${filter.operator} $fieldValue (${fieldValue.javaClass})")
			}
		} catch (e: Exception) {
			throw RuntimeException(
				"booleanFilter $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass}) crashed: ${e.message}",
				e,
			)
		}
	}

	private fun localDateTimeFilter(
		field: Field<LocalDateTime>,
		filter: FilterSpec,
	): Condition {
		val fieldName = StringUtils.toSnakeCase(getPath(filter))
		val fieldValue = filter.getValue<LocalDateTime?>()
		try {
			return if (filter.operator === FilterOperator.EQ) {
				if (fieldValue != null) {
					field.eq(fieldValue)
				} else {
					field.isNull()
				}
			} else if (filter.operator === FilterOperator.NEQ) {
				if (fieldValue != null) {
					field.eq(fieldValue).not()
				} else {
					field.isNotNull()
				}
			} else if (filter.operator === FilterOperator.GT) {
				field.gt(fieldValue)
			} else if (filter.operator === FilterOperator.GE) {
				field.ge(fieldValue)
			} else if (filter.operator === FilterOperator.LT) {
				field.lt(fieldValue)
			} else if (filter.operator === FilterOperator.LE) {
				field.le(fieldValue)
			} else {
				throw IllegalArgumentException("Unsupported LocalDateTime operator $fieldName ${filter.operator} $fieldValue (${fieldValue.javaClass})")
			}
		} catch (e: Exception) {
			throw RuntimeException(
				"localDateTimeFilter $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass}) crashed: ${e.message}",
				e,
			)
		}
	}

	private fun offsetDateTimeFilter(
		field: Field<OffsetDateTime>,
		filter: FilterSpec,
	): Condition {
		val fieldName = StringUtils.toSnakeCase(getPath(filter))
		var fieldValue = filter.getValue<OffsetDateTime?>()
		try {
			return if (fieldValue == null) {
				if (filter.operator === FilterOperator.EQ) {
					field.isNull()
				} else if (filter.operator === FilterOperator.NEQ) {
					field.isNotNull()
				} else {
					throw IllegalArgumentException("Unsupported OffsetDateTime null operator $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass})")
				}
			} else {
				val zoneOffset = ZoneId.systemDefault().rules.getOffset(Instant.now())
				fieldValue = fieldValue.atZoneSameInstant(zoneOffset).toOffsetDateTime()
				if (filter.operator === FilterOperator.EQ) {
					field.eq(fieldValue)
				} else if (filter.operator === FilterOperator.NEQ) {
					field.eq(fieldValue).not()
				} else if (filter.operator === FilterOperator.GT) {
					field.gt(fieldValue)
				} else if (filter.operator === FilterOperator.GE) {
					field.ge(fieldValue)
				} else if (filter.operator === FilterOperator.LT) {
					field.lt(fieldValue)
				} else if (filter.operator === FilterOperator.LE) {
					field.le(fieldValue)
				} else {
					throw IllegalArgumentException("Unsupported OffsetDateTime operator $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass})")
				}
			}
		} catch (e: Exception) {
			throw RuntimeException(
				"offsetDateTimeFilter $fieldName ${filter.operator} $fieldValue (${fieldValue?.javaClass}) crashed: ${e.message}",
				e,
			)
		}
	}

	@Suppress("UNCHECKED_CAST")
	private fun filter(
		table: Table<*>,
		idField: Field<Int>,
		filter: FilterSpec,
	): Condition? {
		try {
			val fieldName = StringUtils.toSnakeCase(getPath(filter))
			if ("is_closed" == fieldName) {
				return closedFilter(table, filter)
			} else if ("search_text" == fieldName) {
				return searchConditionProvider!!.apply(idField, filter)
			}
			val field = table.field(fieldName)!!
			return if (field.getType() == Integer::class.java) {
				integerFilter(field as Field<Int>, filter)
			} else if (field.getType() == java.lang.String::class.java) {
				stringFilter(field as Field<String>, filter)
			} else if (field.getType() == java.lang.Boolean::class.java) {
				booleanFilter(field as Field<Boolean>, filter)
			} else if (field.getType() == LocalDateTime::class.java) {
				localDateTimeFilter(field as Field<LocalDateTime>, filter)
			} else if (field.getType() == OffsetDateTime::class.java) {
				offsetDateTimeFilter(field as Field<OffsetDateTime>, filter)
			} else {
				throw IllegalArgumentException("Unsupported field type " + fieldName + ": " + field.getType())
			}
		} catch (e: Exception) {
			throw RuntimeException("filter($table.$idField, $filter) crashed ${e.message}", e)
		}
	}

	interface SearchConditionProvider {

		fun apply(
			idField: Field<Int>,
			filter: FilterSpec,
		): Condition?

	}

}
